Saros/I does not end the running session when the shared project is closed

Saros/I does not end the running session when the shared project is closed. This leads to issues as it then tries to apply activities to the disposed project/its modules.

This issue can be solved by implementing a project-closed listener which ends the session if the shared project is closed. The only open issue with this approach is how to synchronously end the session or filter remaining activities to guarantee that no activities are applied to disposed resources. This was further discussed in #539 and #568.

@srossbach Could you explain to me again why it is a problem to stop the session on the event dispacher thread? Also, is there any documentation on this (or was there some documentation on the old website that did not get migrated)?

Either way, independently of whether I stop the session synchronously on the EDT or not, I still get issues when closing the project while Saros is still processing local activities (i.e. typing something in a shared editor and then immediately closing the editor). It leads to an interrupted exception (see stacktrace below) and I have no idea what to do about it. I tried different (hacky) solutions to avoid this (filtering incoming activities, manually shutting down the local IDE listeners, manually stopping the `ActivityHandler`), but it didn't help. But I am not sure I fully understand the problem, so these approaches might have been inappropriate anyways.

Any ideas on what the issues is and/or to do?
```
ERROR 15:30:23,407 [dpp-activity-dispatcher] (ThreadUtils.java:37) Internal Error:
com.intellij.openapi.progress.ProcessCanceledException: java.lang.InterruptedException
	at com.intellij.util.concurrency.Semaphore.waitFor(Semaphore.java:76)
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:173)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:680)
	at saros.intellij.runtime.IntelliJSynchronizer.exec(IntelliJSynchronizer.java:35)
	at saros.intellij.runtime.IntelliJSynchronizer.syncExec(IntelliJSynchronizer.java:21)
	at saros.session.internal.ActivityHandler.dispatchAndExecuteActivities(ActivityHandler.java:363)
	at saros.session.internal.ActivityHandler.access$200(ActivityHandler.java:33)
	at saros.session.internal.ActivityHandler$1.run(ActivityHandler.java:100)
	at saros.util.ThreadUtils$1.run(ThreadUtils.java:35)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:998)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at com.intellij.util.concurrency.Semaphore.waitForUnsafe(Semaphore.java:81)
	at com.intellij.util.concurrency.Semaphore.waitFor(Semaphore.java:73)
	... 9 more
ERROR 15:30:23,408 [dpp-activity-dispatcher] (ThreadUtils.java:38) Original caller:
StackTrace:
	at saros.util.ThreadUtils.wrapSafe(ThreadUtils.java:29)
	at saros.util.ThreadUtils.runSafeAsync(ThreadUtils.java:58)
	at saros.session.internal.ActivityHandler.start(ActivityHandler.java:242)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.doLifecycleMethod(StartableLifecycleStrategy.java:101)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.startComponent(StartableLifecycleStrategy.java:96)
	at saros.repackaged.picocontainer.lifecycle.StartableLifecycleStrategy.start(StartableLifecycleStrategy.java:87)
	at saros.repackaged.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.start(AbstractInjectionFactory.java:84)
	at saros.repackaged.picocontainer.behaviors.AbstractBehavior.start(AbstractBehavior.java:169)
	at saros.repackaged.picocontainer.behaviors.Stored$RealComponentLifecycle.start(Stored.java:132)
	at saros.repackaged.picocontainer.behaviors.Stored.start(Stored.java:110)
	at saros.repackaged.picocontainer.DefaultPicoContainer.potentiallyStartAdapter(DefaultPicoContainer.java:1004)
	at saros.repackaged.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:997)
	at saros.repackaged.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:755)
	at saros.session.internal.SarosSession.start(SarosSession.java:614)
	at saros.negotiation.IncomingSessionNegotiation.startSession(IncomingSessionNegotiation.java:307)
	at saros.negotiation.IncomingSessionNegotiation.accept(IncomingSessionNegotiation.java:174)
	at saros.intellij.ui.wizards.JoinSessionWizard$2.run(JoinSessionWizard.java:94)
	at saros.intellij.ui.wizards.Wizard$2.run(Wizard.java:182)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:731)
	at com.intellij.openapi.progress.impl.CoreProgressManager$5.run(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:586)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:86)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$null$10(ApplicationImpl.java:594)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
DEBUG 15:30:23,409 [AWT-EventQueue-0] (SarosSessionManager.java:337) terminating all running negotiations
DEBUG 15:30:23,411 [AWT-EventQueue-0] (XStreamExtensionProvider.java:192) Unregistering (resetting) converter of class saros.misc.xstream.UserConverter
DEBUG 15:30:23,412 [AWT-EventQueue-0] (XStreamExtensionProvider.java:152) Unregistering (resetting) converter of class saros.misc.xstream.SPathConverter
DEBUG 15:30:23,415 [AWT-EventQueue-0] (AbstractLocalDocumentModificationHandler.java:88) Stopped listening for document events
DEBUG 15:30:23,416 [AWT-EventQueue-0] (AbstractLocalDocumentModificationHandler.java:88) Stopped listening for document events
DEBUG 15:30:23,419 [BinaryChannel-intellij_test_1] (BinaryChannelConnection.java:96) [mode=SOCKS5 (D), id=saros-main-session] intellij_test_1@saros-con.imp.fu-berlin.de/Saros connection closed locally: Socket closed
DEBUG 15:30:23,420 [AWT-EventQueue-0] (DataTransferManager.java:349) closed connection [pool id=saros-main-session:out:intellij_test_1@saros-con.imp.fu-berlin.de/Saros]: [mode=SOCKS5 (D), id=saros-main-session] intellij_test_1@saros-con.imp.fu-berlin.de/Saros
```

--- a/core/src/saros/session/ISarosSession.java
+++ b/core/src/saros/session/ISarosSession.java
@@ -152,6 +152,23 @@ public interface ISarosSession {
   public void removeListener(ISessionListener listener);
 
   /**
+   * Enables or disables the execution of received activities for the given project. If the
+   * execution is disabled, activities for resources of the given projects will be dropped without
+   * being applied.
+   *
+   * <p>This method can be used to disable the execution of received activities for projects that
+   * are no longer available (i.e. are no longer part of the session or are no longer
+   * present/accessible locally).
+   *
+   * @param project the shared project to enable or disable the activity execution for
+   * @param enabled <code>true</code> to enable or <code>false</code> to disable the activity
+   *     execution
+   */
+  default void setActivityExecution(IProject project, boolean enabled) {
+    // NOP
+  }
+
+  /**
    * @return the shared projects associated with this session, never <code>null</code> but may be
    *     empty
    */

--- a/core/src/saros/session/internal/SarosSession.java
+++ b/core/src/saros/session/internal/SarosSession.java
@@ -29,6 +29,7 @@ import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CopyOnWriteArraySet;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.log4j.Logger;
 import saros.activities.FileActivity;
@@ -38,6 +39,7 @@ import saros.activities.IActivity;
 import saros.activities.IFileSystemModificationActivity;
 import saros.activities.IResourceActivity;
 import saros.activities.NOPActivity;
+import saros.activities.SPath;
 import saros.communication.extensions.KickUserExtension;
 import saros.communication.extensions.LeaveSessionExtension;
 import saros.concurrent.management.ConcurrentDocumentClient;
@@ -98,6 +100,8 @@ public final class SarosSession implements ISarosSession {
 
   private final CopyOnWriteArrayList<IActivityProducer> activityProducers =
       new CopyOnWriteArrayList<IActivityProducer>();
+
+  private final Set<IProject> filteredProjects = new CopyOnWriteArraySet<>();
 
   private final List<IActivityConsumer> activeActivityConsumers =
       new CopyOnWriteArrayList<IActivityConsumer>();
@@ -164,6 +168,17 @@ public final class SarosSession implements ISarosSession {
 
         @Override
         public void execute(IActivity activity) {
+          // Filters out resource activities for projects whose activity execution is disabled
+          if (activity instanceof IResourceActivity) {
+            SPath path = ((IResourceActivity) activity).getPath();
+
+            if (path != null && filteredProjects.contains(path.getProject())) {
+              log.debug("Dropped activity for resource of filtered project: " + activity);
+
+              return;
+            }
+          }
+
           /**
            * @JTourBusStop 10, Activity sending, Local Execution, first dispatch:
            *
@@ -557,6 +572,17 @@ public final class SarosSession implements ISarosSession {
   @Override
   public void removeListener(ISessionListener listener) {
     listenerDispatch.remove(listener);
+  }
+
+  @Override
+  public void setActivityExecution(IProject project, boolean enabled) {
+    if (project != null) {
+      if (!enabled) {
+        filteredProjects.add(project);
+      } else {
+        filteredProjects.remove(project);
+      }
+    }
   }
 
   @Override

--- a/intellij/src/saros/intellij/context/SarosIntellijSessionContextFactory.java
+++ b/intellij/src/saros/intellij/context/SarosIntellijSessionContextFactory.java
@@ -6,6 +6,7 @@ import saros.intellij.editor.SelectedEditorStateSnapshotFactory;
 import saros.intellij.editor.annotations.AnnotationManager;
 import saros.intellij.eventhandler.ApplicationEventHandlersFactory;
 import saros.intellij.eventhandler.ProjectEventHandlersFactory;
+import saros.intellij.eventhandler.project.ProjectClosedHandler;
 import saros.intellij.followmode.FollowModeNotificationDispatcher;
 import saros.intellij.project.SharedResourcesManager;
 import saros.intellij.project.filesystem.ModuleInitialization;
@@ -23,6 +24,9 @@ public class SarosIntellijSessionContextFactory extends SarosCoreSessionContextF
     container.addComponent(SharedIDEContext.class);
     container.addComponent(ApplicationEventHandlersFactory.class);
     container.addComponent(ProjectEventHandlersFactory.class);
+
+    // Project interaction
+    container.addComponent(ProjectClosedHandler.class);
 
     // Editor interaction
     container.addComponent(LocalEditorHandler.class);

--- a/intellij/src/saros/intellij/eventhandler/project/ProjectClosedHandler.java
+++ b/intellij/src/saros/intellij/eventhandler/project/ProjectClosedHandler.java
@@ -1,0 +1,107 @@
+package saros.intellij.eventhandler.project;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.util.messages.MessageBusConnection;
+import org.jetbrains.annotations.NotNull;
+import saros.filesystem.IProject;
+import saros.intellij.context.SharedIDEContext;
+import saros.intellij.filesystem.IntelliJProjectImpl;
+import saros.repackaged.picocontainer.Disposable;
+import saros.session.ISarosSession;
+import saros.session.ISarosSessionManager;
+import saros.session.SessionEndReason;
+
+/**
+ * Ends the current session when a project containing shared modules is closed and disables the
+ * activity execution for resources of the closed project. This is done to avoid calls on disposed
+ * project objects.
+ *
+ * <p>The handler is enabled by default after instantiation.
+ */
+public class ProjectClosedHandler implements Disposable {
+
+  private final ISarosSessionManager sarosSessionManager;
+  private final ISarosSession sarosSession;
+
+  private final MessageBusConnection messageBusConnection;
+
+  @SuppressWarnings("FieldCanBeLocal")
+  private final ProjectManagerListener projectManagerListener =
+      new ProjectManagerListener() {
+        @Override
+        public void projectClosing(@NotNull Project project) {
+          Project sessionProject = sarosSession.getComponent(SharedIDEContext.class).getProject();
+
+          if (sessionProject.equals(project)) {
+            disableActivityExecutionForProjectModules(project);
+
+            Thread shutdownThread =
+                new Thread(
+                    () -> {
+                      assert !ApplicationManager.getApplication().isDispatchThread()
+                          : "The session must not be shut down from the EDT";
+
+                      sarosSessionManager.stopSession(SessionEndReason.LOCAL_USER_LEFT);
+                    });
+
+            shutdownThread.start();
+          }
+        }
+
+        /**
+         * Disables the activity execution for the modules of the given project.
+         *
+         * <p>This method is used to report the modules of a closed project to the session in order
+         * to prevent it from trying to run any activities on disposed resources.
+         *
+         * @param project the project whose modules to disable the activity execution for
+         * @see ISarosSession#setActivityExecution(IProject,boolean)
+         */
+        private void disableActivityExecutionForProjectModules(@NotNull Project project) {
+          for (Module module : ModuleManager.getInstance(project).getModules()) {
+            IProject wrappedModule;
+
+            try {
+              wrappedModule = new IntelliJProjectImpl(module);
+
+            } catch (IllegalArgumentException | IllegalStateException exception) {
+              continue;
+            }
+
+            if (sarosSession.isShared(wrappedModule)) {
+              sarosSession.setActivityExecution(wrappedModule, false);
+            }
+          }
+        }
+      };
+
+  /**
+   * Instantiates the <code>ProjectClosedHandler</code>. Registers the held <code>
+   * ProjectManagerListener</code> with the application <code>MessageBus</code>.
+   *
+   * @param sarosSessionManager the <code>SarosSessionManager</code> instance
+   * @param sarosSession the current <code>SarosSession</code>
+   * @see MessageBusConnection
+   * @see ProjectManagerListener
+   */
+  public ProjectClosedHandler(
+      ISarosSessionManager sarosSessionManager, ISarosSession sarosSession) {
+
+    this.sarosSessionManager = sarosSessionManager;
+    this.sarosSession = sarosSession;
+
+    messageBusConnection = ApplicationManager.getApplication().getMessageBus().connect();
+    messageBusConnection.subscribe(ProjectManager.TOPIC, projectManagerListener);
+  }
+
+  /** Disconnects from the message bus when the plugin context is disposed. */
+  @Override
+  public void dispose() {
+    messageBusConnection.disconnect();
+  }
+}

--- a/intellij/src/saros/intellij/runtime/IntelliJSynchronizer.java
+++ b/intellij/src/saros/intellij/runtime/IntelliJSynchronizer.java
@@ -3,6 +3,8 @@ package saros.intellij.runtime;
 import com.intellij.openapi.application.Application;
 import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import org.apache.log4j.Logger;
 import saros.synchronize.UISynchronizer;
 
 /**
@@ -10,6 +12,8 @@ import saros.synchronize.UISynchronizer;
  * {@link Application#invokeAndWait(Runnable, ModalityState)}.
  */
 public class IntelliJSynchronizer implements UISynchronizer {
+
+  private static final Logger log = Logger.getLogger(IntelliJSynchronizer.class);
 
   @Override
   public void asyncExec(Runnable runnable) {
@@ -32,7 +36,11 @@ public class IntelliJSynchronizer implements UISynchronizer {
     if (async) {
       application.invokeLater(runnable, ModalityState.defaultModalityState());
     } else {
-      application.invokeAndWait(runnable, ModalityState.defaultModalityState());
+      try {
+        application.invokeAndWait(runnable, ModalityState.defaultModalityState());
+      } catch (ProcessCanceledException e) {
+        log.error("Synchronous execution on EDT interrupted - " + runnable, e);
+      }
     }
   }
 }

